Update operators.rst: Improved documentation for the context parameter

shivannakarthik opened the pull request:
Improved documentation for the context parameter in templated callables, including a clearer explanation of available context variables and their usage. Expanded and refined code examples to better illustrate how to
access runtime information. These changes aim to make the documentation more accessible and
actionable for users customizing operator behavior while leveraging templating features in Airflow.

- Clarified usage of the `context` parameter in callables
- Expanded context variable documentation and examples

closes: #48374

shivannakarthik committed:
Update operators.rst

- Clarified usage of the `context` parameter in callables.
- Expanded context variable documentation and examples.

shivannakarthik committed:
Update operators.rst

- Clarified usage of the `context` parameter in callables.
- Expanded context variable documentation and examples.

ashb commented on airflow-core/docs/core-concepts/operators.rst:
```diff
@@ -84,11 +84,31 @@ Here, ``{{ ds }}`` is a templated variable, and because the ``env`` parameter of
+
 .. code-block:: python
+    from typing import Any, Dict
+    import jinja2
```
This is a bad pattern -- jinja2 is a suprisingly expensive module to import, so we don't want to import it all the time as it will likely slow down dag parsing.

```suggestion
    from typing import TYPE_CHECKING,Any, Dict

    if TYPE_CHECKING:
        import jinja2
```

shivannakarthik replied:
Thanks @ashb for the review comment. Have incorporated the changes requested and removed Any, Dict as its no longer required.

r-richmond commented on airflow-core/docs/core-concepts/operators.rst:
```diff
@@ -84,11 +84,31 @@ Here, ``{{ ds }}`` is a templated variable, and because the ``env`` parameter of
 
-    def build_complex_command(context, jinja_env):
+
+    def build_complex_command(context: Dict[str, Any], jinja_env: jinja2.Environment) -> str:
```
Airflow has a typed dict 'Context' now so you can replace the arbitrary dict[str, Any] with that.

Also it renders your above key examples a little duplicative. I'd update the above just to point to the Context definition so you can access the full list easier.

shivannakarthik replied:
Thanks for the review @r-richmond . I have incorporated the changes and linked the example to doc as well as git, (git and doc may not be in sync all the time)

shivannakarthik committed:
Update operators.rst

Incorporating feedback. Removing specific examples and adding link to the Context docs. Using typed dict for context.

shivannakarthik committed:
Update operators.rst

Incorporating feedback. Removing specific examples and adding link to the Context docs. Using typed dict for context.

shivannakarthik committed:
Update operators.rst

Added additional syntax

shivannakarthik committed:
Update operators.rst

Added additional syntax

uranusjr commented on airflow-core/docs/core-concepts/operators.rst:
```diff
@@ -84,11 +84,25 @@ Here, ``{{ ds }}`` is a templated variable, and because the ``env`` parameter of
 
 You can also pass in a callable instead when Python is more readable than a Jinja template. The callable must accept two named arguments ``context`` and ``jinja_env``:
 
+The ``context`` parameter is Airflow's ``Context`` object that provides access to runtime information about the current task execution. You can access its contents using standard dictionary syntax (``context["key"]``) for direct access , or ``context.get('key', default)`` for safer access with optional defaults. The context contains all the same variables that are available in Jinja templates. Th ...
```
Can we just say Context works like a Python dict? The first half of this paragraph spends a lot of words for a concept that should be basic in Python. If we really want to, we can just link to Python’s documentation on dict interface.

uranusjr commented on airflow-core/docs/core-concepts/operators.rst:
```diff
@@ -84,11 +84,25 @@ Here, ``{{ ds }}`` is a templated variable, and because the ``env`` parameter of
 
+The ``context`` parameter is Airflow's ``Context`` object that provides access to runtime information about the current task execution. You can access its contents using standard dictionary syntax (``context["key"]``) for direct access , or ``context.get('key', default)`` for safer access with optional defaults. The context contains all the same variables that are available in Jinja templates. Th ...
+
+For a complete list of available context variables see :ref:`Templates reference <templates:variables>` and the :class:`airflow.utils.context.Context` type definition on `GitHub <https://github.com/apache/airflow/blob/main/airflow-core/src/airflow/utils/context.py>`_.
```
```suggestion
For a complete list of available context variables see :ref:`Templates reference <templates:variables>`.
```

Please do not link directly to GitHub, especially the main branches. It changes.

uranusjr commented on airflow-core/docs/core-concepts/operators.rst:
```diff
@@ -84,11 +84,25 @@ Here, ``{{ ds }}`` is a templated variable, and because the ``env`` parameter of
+
+    if TYPE_CHECKING:
+        import jinja2
+        from airflow.utils.context import Context
```
```suggestion

        from airflow.sdk import Context
```

shivannakarthik replied:
My understanding was that airflow.utils.context represents internal context related utilities and airflow.sdk provides stable interface for interacting with Airflow's context. 
```from airflow.sdk.definitions.context import Context``` - this is a re-export in airflow.utils.context.  I do see this references in latest code examples.

shivannakarthik committed:
Merge branch 'improve-context-param-doc' of https://github.com/shivannakarthik/airflow into improve-context-param-doc

shivannakarthik committed:
Merge branch 'main' into improve-context-param-doc

shivannakarthik committed:
Update operators.rst

Pushing modifications based on review comments. Removing Git links as they are prone to changes and linking python specific code to its relevant docs on python.

shivannakarthik committed:
Merge branch 'improve-context-param-doc' of https://github.com/shivannakarthik/airflow into improve-context-param-doc

shivannakarthik reviewed (commented):
Have made requested changes. Please review and merge @uranusjr @potiuk

potiuk reviewed (commented):
LlGTM
